File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        low = 1
        high = position[-1]
        
        def isValid(gap):
            placed = 1
            prev = position[0]
            for p in range(1, len(position)):
                if(position[p] - prev >= gap):
                    prev = position[p]
                    placed += 1
                    if(placed == m):
                        return True
            return False
                    
        ans = 0
        while(low <= high):
            mid = (high - low) // 2 + low
            if(isValid(mid)):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1
        
        return ans
                

File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.go
func maxDistance(position []int, m int) int {
    
    var isPossible = func(val int) bool {
        var prev = position[0]
        var ballsPlaced = 1
        for i := 1; i < len(position); i++ {
            if position[i] - prev >= val {
                prev = position[i]
                ballsPlaced += 1
            }
        }
        
        return ballsPlaced >= m
    }
    
    slices.Sort(position)
    var low, high = 0, 0
    for _, p := range position {
        high = max(high, p)
    }
    
    var res = 0
    for low <= high {
        var mid = (low + high) >> 1
        var possible = isPossible(mid)
        if possible {
            res = mid
            low = mid + 1
        } else {
            high = mid - 1
        }
    }
    
    return res
}